crawl-ref/source/annotate.cc

```diff
@@ -0,0 +1,240 @@
+
+enum AnnotationPossibilities
+{
+    // Annotate the dungeon floor the player character is currently on
+    ID_HERE     = -100,
+
+    // Annotate one level up
+    ID_UP       = -99,
+
+    // Annotate one level down
+    ID_DOWN     = -101,
+
+    // Cancel the whole thing
+    ID_CANCEL   = -102,
+};
+
+static int _prompt_annotate_branch()
+{
+    int branch = BRANCH_DUNGEON;     // Default
+
+    //XXX these 3 lines make a vector containing all branches.
+    //there must be a cleaner way to do this.
+    vector<branch_type> brs;
+    for (branch_iterator it; it; ++it)
+        brs.push_back(it->id);
+
+    // Don't kill the prompt even if the only branch we know is the main dungeon
+    // This keeps things consistent for the player.
+    if (brs.size() < 1)
+        return branch;
+
+    level_id curr = level_id::current();
+    while (true)
+    {
+        // Make the little overview
+        // (D) Dungeon        (T) Temple         (L) Lair           etc.
+        // at most 4 branches on 1 line
+        clear_messages();
+        int linec = 0;
+        string line;
+        for (branch_type br : brs)
+        {
+            if (linec == 4)
+            {
+                linec = 0;
+                mpr(line);
+                line = "";
+            }
+            line += make_stringf("(%c) %-14s ",
+                                 branches[br].travel_shortcut,
+                                 branches[br].shortname);
+            ++linec;
+        }
+        if (!line.empty())
+            mpr(line);
+
+        // tells you what the default behaviour is
+        mprf(MSGCH_PROMPT, "Annotate which level? (. - %s)",
+             //curr.c_str()); // TODO Fix this line, and replace the below
+             "current level");
+
+        int keyin = get_ch();
+        switch (keyin)
+        {
+        CASE_ESCAPE           // TODO test whether these 2 lines are necessary
+            return ID_CANCEL; //
+        // special keys are allowed in analogy to shift-G,
+        // but the player isn't told explicitly.
+        // current special cases are Enter, ., <, >, and control-P
+        // for current level, level 1 up, level 1 down, and the parent branch.
+        case '\n': case '\r': case '.':
+            return ID_HERE;
+        case '<':
+            return ID_UP;
+        case '>':
+            return ID_DOWN;
+        case CONTROL('P'):
+            {
+                const branch_type parent = parent_branch(curr.branch);
+                if (parent < NUM_BRANCHES)
+                    return parent;
+            }
+            break;
+        default:
+            // Is this a branch hotkey?
+            for (branch_type br : brs)
+            {
+                if (toupper_safe(keyin) == branches[br].travel_shortcut)
+                    // It is possible to annotate a not-generated branch.
+                    return br;
+            }
+            // Otherwise cancel
+            return ID_CANCEL;
+        }
+    }
+}
```

Comment:
I belief it is impossible to exit the switch statement without going for one of the cases, because I included "default". Though I just noticed that _case control(P)_ has the possibility of ending with **break** rather than **return**, so I do indeed need to fix something.